### ISSUE TITLE
conditional generator sample_zero 

### ISSUE THREAD

**Issue by oneiroid** [pbmartins/ctgan-tf#17]
Do I understand it right that conditional generator saves categorical part of the training dataset (complete, all the rows) in dump, and uses it when generating synthetic data?

**Comment by pbmartins** [comment]
> Do I understand it right that conditional generator saves categorical part of the training dataset (complete, all the rows) in dump, and uses it when generating synthetic data?

Yes, it does.
It is a clear flaw of both this and the original PyTorch solution.
Obviously, with the provided test example, this is not an issue, given that it has a relatively small number of samples (~32K). However, if we use a dataset with 10e6 rows and 100 discrete columns, the output file would certainly be around ~7GB.

As this was mainly a PoC-like implementation of the original solution, I did not look into these kind of details (which are absolutely important), specially on non-PyTorch code (the Conditional Generator was mainly NumPy code, so i left most of it intact, just changed some variable names to make it pass linting analysis).

Nonetheless, this is something we can look into now.
Right now, the structure being stored in the `ConditionalGenerator._model` variable is a list of 1-D Numpy arrays which contain the index of the One-Hot encoded discrete value for each row of the dataset. Given that `ConditionalGenerator._sample_zero` just randomly samples a row from a given column (i.e., it selects a value from the selected NumPy array in the `ConditionalGenerator._model` list), I guess we could map these 1-D arrays to dict-like structures, in which the keys would correspond to the unique values of each array, and the values to the count of those values.
We could use something like this (bear with me, this is not the most optimized solution, but just an overview of what we can do):

```
for col in ConditionalGenerator._model:
    unique_elements, counts_elements = np.unique(col, return_counts=True)
```

Obviously, if every column has exclusively unique values, and we have high number of rows and columns, we would end up on the same sport as we are now. However, this is a more unlikely scenario.

Nonetheless, thank you for the heads up!

**Comment by oneiroid** [comment]
Wow, thnx a lot for such rapid response and comprehensive answer! Btw, your repo is a cake for real - I learned a lot from it, sending my gratitude'n'restecp. I understand it so that original authors wrote the sample_zero method because they had no time to implement smth valid for synthesis after model is trained... 
They shouldn't have published evaluation metrics though, neither do benchmark comparisons with other synth data gen methods - not mentioning this detail looks to me like plain cheating... Or if I don't understand something?

Another question: I'm trying to train your implementation on a dataset without categoricals whatsoever - and generator loss skyrockets to space, while critic loss does the opposite, and results are terrible :)))
So if you have any idea what is the reason? Might that be because of  gradient penalty specifics - it grows at the same rate as generator's loss.
I've seen this question being directed to original authors - but it seems they are not in a mood for answering for some months now :))

**Comment by pbmartins** [comment]
> Wow, thnx a lot for such rapid response and comprehensive answer! Btw, your repo is a cake for real - I learned a lot from it, sending my gratitude'n'restecp. I understand it so that original authors wrote the sample_zero method because they had no time to implement smth valid for synthesis after model is trained...
> They shouldn't have published evaluation metrics though, neither do benchmark comparisons with other synth data gen methods - not mentioning this detail looks to me like plain cheating... Or if I don't understand something?

Well, given the nature of the `ConditionalGenerator`, I assume there's a need to store the original discrete variables distribution, and that's what they came up with. As we discussed, it is not perfect as it is now, due to memory constraints.

> Another question: I'm trying to train your implementation on a dataset without categoricals whatsoever - and generator loss skyrockets to space, while critic loss does the opposite, and results are terrible :)))
> So if you have any idea what is the reason? Might that be because of gradient penalty specifics - it grows at the same rate as generator's loss.
> I've seen this question being directed to original authors - but it seems they are not in a mood for answering for some months now :))

Well, this is tricky. I'll be blunt, I've only tested the stability of training with discrete variables, but I'm seeing that the losses explode without categorical values. I've quickly tested the approach of using a single dummy discrete value, as described [here](https://github.com/sdv-dev/CTGAN/issues/46), and it seems to provide promissing results, even though I did not conduct a thorough research on it.
I guess this is something to look into too, and an issue that we can open up, and further investigate.

**Comment by oneiroid** [comment]
> > Wow, thnx a lot for such rapid response and comprehensive answer! Btw, your repo is a cake for real - I learned a lot from it, sending my gratitude'n'restecp. I understand it so that original authors wrote the sample_zero method because they had no time to implement smth valid for synthesis after model is trained...
> > They shouldn't have published evaluation metrics though, neither do benchmark comparisons with other synth data gen methods - not mentioning this detail looks to me like plain cheating... Or if I don't understand something?
> 
> Well, given the nature of the `ConditionalGenerator`, I assume there's a need to store the original discrete variables distribution, and that's what they came up with. As we discussed, it is not perfect as it is now, due to memory constraints.

You are perfectly right, I agree.
And don't agree - it's not perfect not because of memory constraints, but because sampling from original dataset is not generation of synthetic data. It's easy to implement conditional sampling properly, which probably would degrade final score a bit. Publishing benchmark comparison with other methods is kinda creepy though, if you know that your score is not final.

> 
> > Another question: I'm trying to train your implementation on a dataset without categoricals whatsoever - and generator loss skyrockets to space, while critic loss does the opposite, and results are terrible :)))
> > So if you have any idea what is the reason? Might that be because of gradient penalty specifics - it grows at the same rate as generator's loss.
> > I've seen this question being directed to original authors - but it seems they are not in a mood for answering for some months now :))
> 
> Well, this is tricky. I'll be blunt, I've only tested the stability of training with discrete variables, but I'm seeing that the losses explode without categorical values. I've quickly tested the approach of using a single dummy discrete value, as described [here](https://github.com/sdv-dev/CTGAN/issues/46), and it seems to provide promissing results, even though I did not conduct a thorough research on it.
> I guess this is something to look into too, and an issue that we can open up, and further investigate.

And we will! 
I'll be checking if we may get rid of discrete vars by embedding them as whole via vae, concatenating back with numercals and train regular wgan... Do you think it can work?

**Comment by pbmartins** [comment]
> > > Wow, thnx a lot for such rapid response and comprehensive answer! Btw, your repo is a cake for real - I learned a lot from it, sending my gratitude'n'restecp. I understand it so that original authors wrote the sample_zero method because they had no time to implement smth valid for synthesis after model is trained...
> > > They shouldn't have published evaluation metrics though, neither do benchmark comparisons with other synth data gen methods - not mentioning this detail looks to me like plain cheating... Or if I don't understand something?
> > 
> > 
> > Well, given the nature of the `ConditionalGenerator`, I assume there's a need to store the original discrete variables distribution, and that's what they came up with. As we discussed, it is not perfect as it is now, due to memory constraints.
> 
> You are perfectly right, I agree.
> And don't agree - it's not perfect not because of memory constraints, but because sampling from original dataset is not generation of synthetic data. It's easy to implement conditional sampling properly, which probably would degrade final score a bit. Publishing benchmark comparison with other methods is kinda creepy though, if you know that your score is not final.

Regarding this, I'll look into it when I have some spare time. But I'm still a bit unsure about the solution that I proposed (extracting the latent distribution of the discrete vars by counting values). What do you think?

> 
> > > Another question: I'm trying to train your implementation on a dataset without categoricals whatsoever - and generator loss skyrockets to space, while critic loss does the opposite, and results are terrible :)))
> > > So if you have any idea what is the reason? Might that be because of gradient penalty specifics - it grows at the same rate as generator's loss.
> > > I've seen this question being directed to original authors - but it seems they are not in a mood for answering for some months now :))
> > 
> > 
> > Well, this is tricky. I'll be blunt, I've only tested the stability of training with discrete variables, but I'm seeing that the losses explode without categorical values. I've quickly tested the approach of using a single dummy discrete value, as described [here](https://github.com/sdv-dev/CTGAN/issues/46), and it seems to provide promissing results, even though I did not conduct a thorough research on it.
> > I guess this is something to look into too, and an issue that we can open up, and further investigate.
> 
> And we will!
> I'll be checking if we may get rid of discrete vars by embedding them as whole via vae, concatenating back with numercals and train regular wgan... Do you think it can work?

I've opened a new issue to discuss this: #19  🙂

**Comment by oneiroid** [comment]
Hey, I've checked. Turns out that we can! :)
I've trained vae on categorical columns only with embedding size 32, Encoded them and merged back to dataset, added 1 dummy cat var and trained your (very cool) implementation.
These are the results (correlations not that good... Year-Month, Agency Number and First Initial are the categoricals. Btw, shall I close this issue? )

![image](https://user-images.githubusercontent.com/3052196/83968530-96833080-a8d2-11ea-98ed-7977bdbaa7fb.png)

![image](https://user-images.githubusercontent.com/3052196/83968554-b7e41c80-a8d2-11ea-894f-f7eb9c567f1d.png)

![image](https://user-images.githubusercontent.com/3052196/83968561-c599a200-a8d2-11ea-924c-d0e19bd2822e.png)

**Comment by pbmartins** [comment]
Hi! Let me just see if I understood this correctly.
So, you replaced the Data Transformation using the Variational GMM with the VAE, or did you pre-process the data using the VAE even before passing it to the whole Synthesizer class?
Maybe I was not explicit enough, but did you:
1 - Data -> CTGANSynthesizer -> DataTransformer (Replaced with VAE) -> GAN
OR
2 - Data -> VAE -> CTGANSynthesizer -> DataTransformer (with VGM and OHE) -> GAN

Nonetheless, you added a dummy categorical value, so that the losses don't explode?

Regarding the correlations between variables, that is something that I've also noticed before.
In my eyes, this model (CTGAN) can be further improved by adding some kind of penalty (like the conditional loss or GP) when the correlations start to disappear. What do you think of this?

Regarding this issue, maybe I'll reference your answer in [here](https://github.com/pbmartins/ctgan-tf/issues/19), right? Then, I'll close that one after tinkering with the implementation (maybe change it to automatically add the dummy categorical column).

**Comment by oneiroid** [comment]
Sorry for late reply, these days are strange and speed by fast. 
In short - approximately 2nd.

In full (by now it is bit different):
Mixed Data -> ohe(cat_cols), minmaxscaler(num_cols) -> VAE (with mmd loss) -> embeddings (64 dim) + dummy_cat_col -> CTGANSynthesizer (not using ctgan transformer on embedding columns)
Surprisingly, VAE reconstructs mixed data with 99% perfection. 

`
Regarding the correlations between variables, that is something that I've also noticed before.
In my eyes, this model (CTGAN) can be further improved by adding some kind of penalty (like the conditional loss or GP) when the correlations start to disappear. What do you think of this?
`
Would be nice to check it out! 
Re the last question - you need not asking *_*

Ps. I'll share the code, don't have  no time now, sorry.
